cpu/cc2538/Makefile.cc2538

```diff
@@ -84,7 +85,7 @@ CUSTOM_RULE_LINK=1
 
 %.elf: $(CPU_STARTFILES) %.co $(PROJECT_OBJECTFILES) $(PROJECT_LIBRARIES) contiki-$(TARGET).a $(LDSCRIPT)
 	$(TRACE_LD)
-	$(Q)$(LD) $(LDFLAGS) ${filter-out $(LDSCRIPT) %.a,$^} ${filter %.a,$^} $(TARGET_LIBFILES) -o $@
+	$(Q)$(LD) $(LDFLAGS) -Wl,--start-group ${filter-out $(LDSCRIPT) %.a,$^} ${filter %.a,$^} $(TARGET_LIBFILES) -lc -Wl,--end-group -o $@
```

Comment:
Yes, I do remember that. `-Wl,--start-group` and `-Wl,--end-group` are meant to be used to enclose a list of archive files to be searched in a loop until all symbols are resolved. So object files have nothing to do in that list.

Normally, ld goes through the objects and archives in the order they are passed. The objects within archives that do not contain definitions unresolved by previously kept objects are discarded (even without using ld garbage collection, which is completely unrelated).

The issue here is that Contiki puts nearly everything into an archive. Because of that, without the archive group trick, `libc.a` is implicitly linked at the end, so `malloc()` and `_sbrk_r()` are kept because the objects and archives previously linked need them (if `malloc()` is used somewhere in an example). But in this process, `_sbrk()` was discarded with your `syscall.o` within `contiki-cc2538dk.a` because nothing directly needed it at the time it was linked (`libc.a` containing `malloc()` needing `_sbrk_r()` needing `_sbrk()` is linked later). Also, `contiki-cc2538dk.a` needs `libc.a` to resolve many other definitions. In the end, we have `contiki-cc2538dk.a` and `libc.a` depending on each other, which is impossible to link unless they are grouped in a symbol resolution loop.

It would be possible to avoid the archive groups, and even passing `-lc`, if `syscall.o` were not included in `contiki-cc2538dk.a`, or even better, if `contiki-cc2538dk.a` were removed (e.g. changed into a big `contiki-cc2538dk.o`). `contiki-cc2538dk.a` is not a real feature-centric library built in order to stay and be linked with applications at any time later, but just a convenient object bag to ease the build process of several examples at once, and in this case it complicates things. Also, being limited to a single instance of a given file name is not clean, but this is another topic.

There are also some magical dangerous tricks used in Contiki that depend on this behavior of ld towards archives. E.g., the commit f7baf5aba2cb2916eb6e44c8ed40bcc70e145cd3 (@g-oikonomou) has introduced weak aliases for the USB and UART ISRs that can only work because of that. With the current code, if UART is not used, then nothing needs `uart.o` within `contiki-cc2538dk.a`, so `uart.o` is fully discarded, even though it contains the strong `uart0_isr()` definition, because ld is satisfied at this point with the weak alias. Instead, with a big `contiki-cc2538dk.o`, the strong definition would win, so the wrong ISR would be linked in, which could be fixed by keeping the strong definition of `uart0_isr()` in `uart.c` only `#if UART_CONF_ENABLE`. It is probably safer not to rely on the current implicit behavior, because changing the build system would introduce the issue described here, and it would remain unnoticed for a very long time. With this example, we also see that the build system builds a lot of things that are using for a given example, which is cool for testing with Travis, but this slows down the build.